ClassScanner - Throw an exception if a class-file doesn't have the expected class

totten opened the pull request:
Overview
----------------------------------------

This improves error reporting when working with scanned classes.

Suppose you create a class-file like this:

```php
<?php
// FILE: Civi/Foo/Bar.php
class Bar {
}
```

It's missing the `namespace`!  Similarly, you might make a typo in the `namespace` or `class`. So what happens with it?

(Note: This is one of two alternative PRs: #24347 does a warning, #24348 does an exception.)

Before
----------------------------------------

The class is silently ignored. Functionality doesn't work. 

Depending on the specific use-case, you may face the implacable silence of software ("Why doesn't my file load?") or you might get some random error (*that has to be traced back to the misnamed class*).

After
----------------------------------------

Throws an error.

Comments
----------------------------------------

I originally/reflexively wrote this as a warning (#24347). But on second thought, there may be a good argument for throwing an exception:

1. The `scan-classes` stuff is fairly new and not widely used yet, so there's not much concern about breaking things that  (somehow/bizarrely) rely on mismatched names.
2. If you intentionally have weird/different class-file structure, then you're not supposed to enable `scan-classes`. (That's the point of making it toggleable....)
3. An exception is more apparent to the developer than a warning in a log.
4. It makes this kind of error tantamount to a syntax error.

totten committed:
ClassScanner - Throw an exception if a scanned file doesn't have an expected class

demeritcowboy commented:
I'd vote for this version.

seamuslee001 commented:
Question does this happen at boot time? if so could it cause problems with boot?

totten commented:
@seamuslee001 Yes, if there's a malformed class, I expect it would fail pretty earlier. For example, #24276 aims to use scanned-classes to help compile the container. If you have a malformed class, then `createContainer()` will fail.

But if you zoom out, this only applies to folders/extensions/classes for which class-scanning has been explicitly enabled (ie `<mixin>scan-classes@1.0.0</mixin>`), and the error represents a structural incongruity.

totten committed:
UnitTests - Tighten class-scanning for `civicrm-core:tests/phpunit`

This addresses a bug where:

1. Class-scanning of `civicrm-core:tests/phpunit/*` works with core-test-suites
2. Class-scanning of `civicrm-core:tests/phpunit/*` fails with civiimport test-suite.

The difference is that core-tests setup class-loading for
`civicrm-core:tests/phpunit/`; `civimport` understandably doesn't.

Class-scanning is only viable if we regard the specific classes a loadable.

This further tightens it:

1. Since core-test-suites can load `civicrm-core:tests/phpunit/*`, they will scan it
2. Since civiimport test-suite cannot load it, it will not scan it.

totten commented:
@demeritcowboy @seamuslee001  - So that first failure revealed that there had been an improper scan going-on... 

With the second commit, it passes now.

demeritcowboy commented on Civi/Core/ClassScanner.php:
```diff
@@ -137,8 +137,10 @@ private static function scanCoreClasses(): array {
     if (\CRM_Utils_Constant::value('CIVICRM_UF') === 'UnitTests') {
-      static::scanFolders($classes, $civicrmRoot . 'tests/phpunit', 'CRM/*/WorkflowMessage', '_');
-      static::scanFolders($classes, $civicrmRoot . 'tests/phpunit', 'Civi/*/WorkflowMessage', '\\');
+      if (strpos(get_include_path(), $civicrmRoot . 'tests/phpunit') !== FALSE) {
```
This strpos might fail on windows but this is only for unit tests so can deal with later if it comes up.
